### DISCUSSION TITLE
How to get CreatedBy name at querying?

### DISCUSSION THREAD

**Discussion by ANMSalman** [jasontaylordev/CleanArchitecture#412] [Q&A] [answered]
Let's say I want to display the name of the user who created a record (Ex: Products, Brands, Customers, etc..), How do I achieve this? Since Identity-related entities are in Infrastructure and my GetAllProductsQuery feature is in the Application layer. How do I add the created user's name to the response object of the GetAllProductsQuery feature?

**Comment by MichaelHochriegl** [comment, marked as answer]
Did you look at the template provided here? There is a [ICurrentUserService.cs](https://github.com/jasontaylordev/CleanArchitecture/blob/main/src/Application/Common/Interfaces/ICurrentUserService.cs). This is used in the [SaveChangesAsync](https://github.com/jasontaylordev/CleanArchitecture/blob/main/src/Infrastructure/Persistence/ApplicationDbContext.cs) method in infrastructure, so that an entity that extends from [AuditableEntity](https://github.com/jasontaylordev/CleanArchitecture/blob/main/src/Domain/Common/AuditableEntity.cs) get's the current user assigned.

If you query an entity that is an AuditableEnity you also have access to the CreatedBy and/or LastModifiedBy prop and can use this to display the user that created/modified the entity.

Update: If you need the current user in any other command/query you can simply ask for an ICurrentUserService and for an IIdentityService using dependency injection and can access the user info with this.

**Reply by ANMSalman** [reply]
This was not I was looking for. Please read the question again. I don't need the current user details. Also IAuditableEntity only has the ID. I want the Name. Not the Id. I want created user's **Name** when querying records

**Reply by MichaelHochriegl** [reply]
I did give you the right answer. You can get the actual user info by taking the UserID returned from the AuditableEntity props by injecting the IIdentityService and calling the GetUserNameAsync. The first paragraph of my answer was to describe the overall setup of this, so that you see how the fields from AuditableEntity are actually set.

Here is a quick mockup of what I ment (Note: there is no errorhandling here, so you would still need to add checks for null and what not):
```C#
using System;
using System.Threading;
using System.Threading.Tasks;
using AutoMapper;
using CleanArchitecture.Application.Common.Interfaces;
using CleanArchitecture.Application.Common.Mappings;
using CleanArchitecture.Domain.Entities;
using MediatR;

namespace CleanArchitecture.Application.TodoLists.Queries.GetTodoItemWithAuthor
{
    public class GetTodoItemWithAuthorQuery : IRequest<TodoItemVm>
    {
        public int Id { get; set; }
    }

    public class GetTodoItemWithAuthorQueryHandler : IRequestHandler<GetTodoItemWithAuthorQuery, TodoItemVm>
    {
        private readonly IApplicationDbContext _context;
        private readonly IMapper _mapper;
        private readonly IIdentityService _identityService;

        public GetTodoItemWithAuthorQueryHandler(IApplicationDbContext context, IMapper mapper, IIdentityService identityService)
        {
            _context = context;
            _mapper = mapper;
            _identityService = identityService;
        }
        
        public async Task<TodoItemVm> Handle(GetTodoItemWithAuthorQuery request, CancellationToken cancellationToken)
        {
            var entry = await _context.TodoItems.FindAsync(request.Id);

            var username = await _identityService.GetUserNameAsync(entry.CreatedBy);
            
            var result = new TodoItemVm
            {
                Title = entry.Title,
                Note = entry.Note,
                Created = entry.Created,
                CreatedById = entry.CreatedBy,
                CreatedByUsername = username
            };

            return result;
        }
    }

    public class TodoItemVm : IMapFrom<TodoItem>
    {
        public string Title { get; set; }
        public string Note { get; set; }
        public DateTime Created { get; set; }

        public string CreatedById { get; set; }
        public string CreatedByUsername { get; set; }
    }
}
```

**Reply by ANMSalman** [reply]
Thanks

**Reply by ali-gol** [reply]
@MichaelHochriegl  What if you want to query list of dataset with CreatedBy user names?
I would like to query 100 data with the related usernames?

**Reply by MichaelHochriegl** [reply]
You could project it by using almost the same approach as I posted above. You would fetch all the `TodoItem`s you want to return and populate the above `TodoItemVm` with the username that you still would get from `GetUserNameAsync`. 

Quick example:
```csharp
var result = entries.Select(async x => new TodoItemVm
{
    Title = x.Title,
    Note = x.Note,
    Created = x.Created,
    CreatedById = x.CreatedBy,
    CreatedByUsername = await _identityService.GetUserNameAsync(x.CreatedBy)
});
```

But I'm not really a fan of this approach.
I would instead argue that if your DB model is right if you find yourself in this situation. As it stands with the example here in this CleanArchitecture repo the user is stored separately from the "main" data. This is fine, but I suggest that if this is requirement for you than maybe think about your DB design instead.
